xclim/core/units.py

```diff
@@ -461,6 +484,74 @@ def to_agg_units(
     return out
 
 
+def _rate_and_amount_converter(
+    da: xr.DataArray, dim: str = "time", to: str = "amount", out_units: str = None
+) -> xr.DataArray:
+    """Private function performing the actual conversion for :py:func:`rate2amount` and :py:func:`amount2rate`."""
+    m = 1
+    u = None  # Default to assume a non-uniform axis
+    label = "lower"
+    time = da[dim]
+
+    try:
+        freq = xr.infer_freq(da[dim])
+    except ValueError:
+        freq = None
+    if freq is not None:
+        multi, base, start_str, _ = parse_offset(freq)
+        if base in ["M", "Q", "A"]:
+            start = time.indexes[dim][0]
+            if start_str != "S":
+                # Anchor is on the end of the period, substract 1 period.
+                start = start - xr.coding.cftime_offsets.to_offset(freq)
+                # In the diff below, assign to upper label!
+                label = "upper"
+            # We generate "time" with an extra element, so we do not need to repeat the last element below.
+            time = xr.DataArray(
+                date_range(
+                    start, periods=len(time) + 1, freq=freq, calendar=get_calendar(time)
+                ),
+                dims=(dim,),
+                name=dim,
+                attrs=da[dim].attrs,
+            )
+        else:
+            m, u = int(multi or "1"), FREQ_UNITS[base]
+
+    # Freq is month, season or year, which are not constant units, or simply freq is not inferrable.
+    if u is None:
+        # Get sampling period lengths in nanoseconds
+        # In the case with no freq, last period as the same length as the one before.
+        # In the case with freq in M, Q, A, this has been dealt with above in `time` and `label` has been update accordingly.
+        dt = (
+            time.diff(dim, label=label)
+            .reindex({dim: da[dim]}, method="ffill")
+            .astype(float)
+        )
+        dt = dt / 1e9  # Convert to seconds
+
+        if to == "amount":
+            tu = (str2pint(da.units) * str2pint("s")).to_reduced_units()
+            out = da * dt * tu.m
+        elif to == "rate":
```

Comment:
You have an if/elif here and a if/else farther down to case switch on `to`. If there are only two options, it could be a bool. If you want to keep the str argument for clarity, then I suggest using if/elif/else.